include/cuco/static_reduction_map.cuh

```diff
@@ -0,0 +1,1131 @@
+ * operations are accessed through non-owning, trivially copyable "view" types:
+ * `device_view` and `mutable_device_view`. The `device_view` class is an
+ * immutable view that allows only non-modifying operations such as `find` or
+ * `contains`. The `mutable_device_view` class only allows `insert` operations.
+ * The two types are separate to prevent erroneous concurrent insert/find
+ * operations.
+ *
+ *  Example:
+ *  \code{.cpp}
+ *
+ * // Empty slots are represented by reserved "sentinel" values. These values should be selected
+ * such
+ * // that they never occur in your input data.
+ * int const empty_key_sentinel = -1;
+ *
+ * // Number of key/value pairs to be inserted
+ * std::size_t const num_elems = 256;
+ *
+ * // average number of values per distinct key
+ * std::size_t const multiplicity = 4;
+ *
+ * // Compute capacity based on a 50% load factor
+ * auto const load_factor     = 0.5;
+ * std::size_t const capacity = std::ceil(num_elems / load_factor);
+ *
+ * // Constructs a map each key with "capacity" slots using -1 as the
+ * // empty key sentinel. The initial payload value for empty slots is determined by the identity of
+ * // the reduction operation. By using the `reduce_add` operation, all values associated with a
+ * // given key will be summed.
+ * cuco::static_reduction_map<cuco::reduce_add<int>, int, int> map{capacity, empty_key_sentinel};
+ *
+ * // Create a sequence of random keys
+ * thrust::device_vector<int> insert_keys(num_elems);
+ * thrust::transform(thrust::device,
+ *                   thrust::make_counting_iterator<std::size_t>(0),
+ *                   thrust::make_counting_iterator(insert_keys.size()),
+ *                   insert_keys.begin(),
+ *                   [=] __device__(auto i) {
+ *                     thrust::default_random_engine rng;
+ *                     thrust::uniform_int_distribution<int> dist(
+ *                       int{1}, static_cast<int>(num_elems / multiplicity));
+ *                     rng.discard(i);
+ *                     return dist(rng);
+ *                   });
+ *
+ * // Insert each key with a payload of `1` to count the number of times each key was inserted by
+ * // using the `reduce_add` op
+ * auto zipped = thrust::make_zip_iterator(
+ *   thrust::make_tuple(insert_keys.begin(), thrust::make_constant_iterator(1)));
+ *
+ * // Inserts all pairs into the map, accumulating the payloads with the `reduce_add` operation
+ * map.insert(zipped, zipped + insert_keys.size());
+ *
+ * // Get a `device_view` and passes it to a kernel where threads may perform
+ * // `find/contains` lookups
+ * kernel<<<...>>>(m.get_device_view());
+ * \endcode
+ *
+ *
+ * @tparam Key Arithmetic type used for key
+ * @tparam Value Type of the mapped values
+ * @tparam Scope The scope in which insert/find operations will be performed by
+ * individual threads.
+ * @tparam Allocator Type of allocator used for device storage
+ */
+template <typename ReductionOp,
```

Comment:
We should do something to enforce requirements on `ReductionOp`. Basically, it needs to be one of the operators provided in `reduction_ops.cuh` or if a custom operation, needs to use `custom_op`.

Reply:
How about a member tag aka an empty struct `cuco::tags::reduction_op`?

Reply:
Here's what I was thinking. A person has 3 options for the `ReductionOp`

1. Use one of the provided `cuco::reduce_*` types.
   - No additional work should be required. Partial specialization could/should remove the `ReductionOp` argument from the constructor
2. Provide a unsynchronized binary callable `T F(T, T)` and `Identity` value
  - This needs to be wrapped by `custom_op` to apply `F` in a CAS loop
  - Ideally we could detect this kind of callable and implicitly wrap it in `custom_op`
3. Provide a synchronized binary callable `T F(atomic_ref<T, Scope>, T)` and `Identity` value
  - User responsible for correct synchronization through `atomic_ref`

Examples:
```
// 1.
// no need to provide `reduce_add{}` 
// No need to provide identity value
cuco::static_reduction_map<cuco::reduce_add<int>, int, int> add_map{capacity, empty_key, alloc}; 

// 2. Unsynchronized binary callable must be wrapped in `custom_op`
struct unsync_add{ 
   int identity = 0; // Must provide identity value
   int operator()(int a, int b){ return a + b; }
};

// internally should wrap `unsync_add` in `custom_op`
cuco::static_reduction_map<unsync_add, int, int> custom_unsync_add_map(capacity, empty_key, unsync_add{}, alloc);

// 3.
stuct sync_add{
   int identity = 0; // Must provide identity value
   template <thread_scope Scope>
   int operator()(atomic_ref<int, Scope> a, int b){ return a.fetch_add(b, memory_order_relaxed); }
};

cuco::static_reduction_map<sync_add, int, int> custom_sync_add_map(capacity, empty_key, sync_add{}, alloc);
```

Reply:
1 & 3 could effectively be merged.

One thing that occurred to me is that the identity value need not be known statically. Not sure what kind of binop would have a runtime determined identity value, but who knows?

Reply:
> How about a member tag aka an empty struct cuco::tags::reduction_op?

@sleeepyjack I'd prefer to create a base class: https://godbolt.org/z/6KqenYenT

Reply:
> How about a member tag aka an empty struct cuco::tags::reduction_op?

@sleeepyjack I'd prefer to create a base class: https://godbolt.org/z/6KqenYenT

Reply:
Re-examining this question again:

I am a bit puzzled about how to distinguish between cases 1) and 3) as it involves extracting the type of the first argument of the `operator()` and check if it is `atomic_ref<T>` or just `T` 🙃. Maybe something like this?:
```
template <typename> struct first_arg;

template <typename F, typename A, typename... Args>
struct first_arg<F(A, Args...)>
{
   using type = A;
};

template <typename T>
using first_arg_t = typename first_arg<T>::type;
```
Also, implicitly switching between sync and non-sync implementations may lead to confusion on the user side.

How about defining a common base class for all built-in (synchronizing) functors. If a user passes a functor that doesn't inherit from this base, it is automatically wrapped in e.g. a CAS loop. 
This way we put the user in charge of deciding whether the functor needs synchronization or not.
Additionally, we could use CRTP to add some convenient type checks to the base class.

Let me know what you think.